### PR TITLE
Release 2.1.6

### PR DESCRIPTION
**Pull request by glennmatthews**
# What's Changed

Lots of fixes in this release!

### Security

- [#5319](https://github.com/nautobot/nautobot/issues/5319) - Updated `cryptography` to 42.0.4 due to CVE-2024-26130. This is not a direct dependency so will not auto-update when upgrading. Please be sure to upgrade your local environment.

### Added

- [#5172](https://github.com/nautobot/nautobot/issues/5172) - Added Collapse Capable Side Navbar: Side Navbar is now able to be expanded and collapsed
- [#5172](https://github.com/nautobot/nautobot/issues/5172) - Added Expandable Main Content: The Main Content part of the UI grows as the Side Navbar collapses and shrinks as the Side Navbar expands.
- [#5172](https://github.com/nautobot/nautobot/issues/5172) - Added Better mobile friendly bottom navbar: This update will switch to vertically aligned bottom nav menu items once a certain media query is hit, making for a better mobile experience.
- [#5172](https://github.com/nautobot/nautobot/issues/5172) - Added automatic Side Navbar collapse for mobile devices.  This is based on media query and will trigger at specific width.
- [#5329](https://github.com/nautobot/nautobot/issues/5329) - Added caching of `ChangeLoggedModelsQuery().as_queryset()` to improve performance when saving many objects in a change-logged context.

### Changed

- [#5082](https://github.com/nautobot/nautobot/issues/5082) - Adjusted Edit / Create panels to occupy more page width on medium and large screens.

### Fixed

- [#4106](https://github.com/nautobot/nautobot/issues/4106) - Fixed inefficient query in VirtualMachine create form.
- ~[#4419](https://github.com/nautobot/nautobot/issues/4419) - Re-added `NavMenuButton` back into `nautobot.apps.ui` module.~
- [#5172](https://github.com/nautobot/nautobot/issues/5172) - Fixed Brand Icon mouseover Background: Fix for mouseover effect on the Brand / Icon (was flashing white background vs being transparent) when in dark mode.
- [#5307](https://github.com/nautobot/nautobot/issues/5307) - Fixed Custom Field form field(s) missing from git repository edit form.
- [#5309](https://github.com/nautobot/nautobot/issues/5309) - Fixed `Tenant` UI detail view breadcrumb with invalid `TenantGroup` filter link.
- [#5309](https://github.com/nautobot/nautobot/issues/5309) - Fixed `TenantGroup` UI detail view with invalid "add tenant" button invalid query_params link.
- [#5309](https://github.com/nautobot/nautobot/issues/5309) - Fixed `DeviceForm` invalid `cluster` field query_params.
- [#5309](https://github.com/nautobot/nautobot/issues/5309) - Fixed `PrefixForm` invalid `vlan` and `vlan_group` fields query_params.
- [#5311](https://github.com/nautobot/nautobot/issues/5311) - Fixed dependencies in various migration files.
- [#5332](https://github.com/nautobot/nautobot/issues/5332) - Fixed Docker image missing OS-level dependencies for SSO (SAML) support.
- [#5334](https://github.com/nautobot/nautobot/issues/5334) - Fixed migration from 1.x failing when specific duplicate prefixes are present.
- [#5343](https://github.com/nautobot/nautobot/issues/5343) - Fixed incorrect reference for `device.device_role` on the Rack detail view for non-racked device objects.
- [#5345](https://github.com/nautobot/nautobot/issues/5345) - Fixed intermittent 405 errors when using the Docker image with SAML authentication.
- [#5346](https://github.com/nautobot/nautobot/issues/5346) - Fixed device LLDP view to work when interface names include a space.
- [#5365](https://github.com/nautobot/nautobot/issues/5365) - Fixed `invalidate_max_depth_cache` itself calculating `max_depth` on querysets without tree fields.

### Documentation

- [#4419](https://github.com/nautobot/nautobot/issues/4419) - Added documentation on `nautobot.apps` import locations.
- [#4419](https://github.com/nautobot/nautobot/issues/4419) - Added documentation about the supported public interfaces.
- [#4419](https://github.com/nautobot/nautobot/issues/4419) - Removed some incorrect content from the documentation about nav menu changes that were reverted during 2.0 development.
- [#4511](https://github.com/nautobot/nautobot/issues/4511) - Added documentation on how to correctly implement NautobotUIViewSet with custom views.
- [#5284](https://github.com/nautobot/nautobot/issues/5284) - Added a quick overview of the most used models.
- [#5311](https://github.com/nautobot/nautobot/issues/5311) - Added documentation on writing custom migrations.
- [#5326](https://github.com/nautobot/nautobot/issues/5326) - Fixed simple typo in creating-location-types-and-locations.md.
- [#5330](https://github.com/nautobot/nautobot/issues/5330) - Updated SSO documentation to include a view for presenting SAML metadata.
- [#5345](https://github.com/nautobot/nautobot/issues/5345) - Added a note to the Nautobot installation documentation about the need to do `pip3 install --no-binary=pyuwsgi` in order to have SSL support in `pyuwsgi`.
- [#5345](https://github.com/nautobot/nautobot/issues/5345) - Added a note to the SSO documentation about the need to do `pip3 install --no-binary=lxml` to avoid incompatibilities between `lxml` and `xmlsec` packages.

### CODE REVIEW

File: nautobot/core/testing/forms.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+from django.test import tag, TestCase
+
+from nautobot.core.forms.fields import DynamicModelChoiceMixin
+from nautobot.core.utils.lookup import get_filterset_for_model
+
+
+# TODO(timizuo): All Form Test cases should inherit from FormTestCases
+@tag("unit")
+class FormTestCases:
```
Comment (glennmatthews) [inline nautobot/core/testing/forms.py:9]: We should export this in `nautobot.apps` and document it in the developer docs and release notes.

File: nautobot/docs/release-notes/version-2.1.md [hunk 2]
```diff
@@ -72,6 +72,55 @@ Support for versions of PostgreSQL prior to 12.0 has been removed as these versi
...
+
+### Added
+
+- [#5172](https://github.com/nautobot/nautobot/issues/5172) - Added Collapse Capable Side Navbar: Side Navbar is now able to be expanded and collapsed
+- [#5172](https://github.com/nautobot/nautobot/issues/5172) - Added Expandable Main Content: The Main Content part of the UI grows as the Side Navbar collapses and shrinks as the Side Navbar expands.
+- [#5172](https://github.com/nautobot/nautobot/issues/5172) - Added Better mobile friendly bottom navbar: This update will switch to vertically aligned bottom nav menu items once a certain media query is hit, making for a better mobile experience.
+- [#5172](https://github.com/nautobot/nautobot/issues/5172) - Added automatic Side Navbar collapse for mobile devices.  This is based on media query and will trigger at specific width.
+- [#5329](https://github.com/nautobot/nautobot/issues/5329) - Added caching of `ChangeLoggedModelsQuery().as_queryset()` to improve performance when saving many objects in a change-logged context.
```
Comment (glennmatthews) [inline nautobot/docs/release-notes/version-2.1.md:87]:
```suggestion
- [#5329](https://github.com/nautobot/nautobot/issues/5329) - Added caching of `ChangeLoggedModelsQuery().as_queryset()` to improve performance when saving many objects in a change-logged context.
- [#5361](https://github.com/nautobot/nautobot/issues/5361) - Added `nautobot.core.testing.forms.FormTestCases` base class.
```

File: nautobot/core/testing/forms.py (hunk 1 above)
Comment (glennmatthews) [reply nautobot/core/testing/forms.py:9]: #5382

File: nautobot/docs/release-notes/version-2.1.md (hunk 2 above)
Comment (glennmatthews) [reply nautobot/docs/release-notes/version-2.1.md:87]: #5382

File: nautobot/docs/release-notes/version-2.1.md [hunk 3]
```diff
@@ -72,6 +72,55 @@ Support for versions of PostgreSQL prior to 12.0 has been removed as these versi
...
+
+### Changed
+
+- [#5082](https://github.com/nautobot/nautobot/issues/5082) - Adjusted Edit / Create panels to occupy more page width on medium and large screens.
+
+### Fixed
+
+- [#4106](https://github.com/nautobot/nautobot/issues/4106) - Fixed inefficient query in VirtualMachine create form.
+- [#4419](https://github.com/nautobot/nautobot/issues/4419) - Re-added `NavMenuButton` back into `nautobot.apps.ui` module.
```
Comment (glennmatthews) [inline nautobot/docs/release-notes/version-2.1.md:96]: Leftover from early implementation of #4419; this was already re-added by #4533

```suggestion
```

File: nautobot/docs/development/apps/api/views/nautobotuiviewset.md [hunk 4]
```diff
@@ -145,3 +145,53 @@ For example, for a DetailView template for `YourAppModel`, the template name wil
...
++/- 1.6.0
+    Via [PR #4045](https://github.com/nautobot/nautobot/pull/4045), notes and changelog views provided by mixins have now been moved to this pattern.
+
+Django REST Framework provides the ability to decorate a method on a ViewSet with `@action(detail=True)` to add the method as a view to the ViewSetRouter. This method must return a fully rendered HTML view.
+
+Below is an example of adding a custom action view to an App's ViewSet. A few considerations to keep in mind:
+
+* The method name is the `action` in Django REST Framework terms.
+* The `action` will be used for [template lookup](#template-naming-for-nautobotuiviewset)
+* The `action` will be used for URL naming and construction (`plugins:<app>:<model>_<action>`, `/plugins/<app>/<model>/<uuid>/<action>/`).
```
Comment (HanlinMiao) [inline nautobot/docs/development/apps/api/views/nautobotuiviewset.md:160]: I dont think we changed the app namespace `plugins` yet. But I just want to make sure

File: nautobot/docs/development/core/index.md [hunk 5]
```diff
@@ -69,6 +69,33 @@ Nautobot adheres to the Semantic Versioning ("SemVer") strategy, which gives us
...
+* Documented configuration options
+
+!!! info
+    The public interface may change the function signature but will not break backwards-compatibility between minor & patch versions in accordance with SemVer.
+
+While anything not explicitly stated above should be not be considered scoped to SemVer, the following should be strictly avoided when developing non-core Nautobot Apps and Jobs.
+
+* Functions or methods that start with an underscore (e.g., `_private_func`)
+* Any functions, classes, and modules **not** in `nautobot.apps.*`.
+* Any modules, classes, functions, or feature that are explicitly marked as "internal" or "experimental" in the documentation
```
Comment (HanlinMiao) [inline nautobot/docs/development/core/index.md:96]:
```suggestion
* Any modules, classes, functions, or features that are explicitly marked as "internal" or "experimental" in the documentation
```

**Review by HanlinMiao — APPROVED** [review]

File: nautobot/docs/development/apps/api/views/nautobotuiviewset.md (hunk 4 above)
Comment (glennmatthews) [reply nautobot/docs/development/apps/api/views/nautobotuiviewset.md:160]: Correct, we have not.

**Review by gsnider2195 — APPROVED** [review]

**Pull request merged**